### PR TITLE
Fix garbage collection

### PR DESCRIPTION
**Pull request by akaszynski**
### First Problem: Unnecessary Scalar Array Copying

As pointed out, we're copying scalars within `add_mesh`. When plotting existing scalars with:

```py
import pyvista
mesh = pyvista.Sphere()
mesh['data'] = range(mesh.n_points)
mesh.plot(scalars='data')
```

We actually create a copy of the scalars:
https://github.com/pyvista/pyvista/blob/316a110376d74fd81834316284703cfdd4b74f8c/pyvista/plotting/mapper.py#L253-L263

This is inefficient, but this actually protects us from underlying garbage collection issues.

---

### Second Problem: Garbage Collection of VTK Arrays

When performing a shallow copy, we end up with a VTK array memory leak. This can be reproduced with:
```py
import pyvista
mesh = pyvista.Sphere()
mesh.point_data['data'] = range(mesh.n_points)
mesh.point_data['data'] = mesh.point_data['data']
```

This issue is caused by us recreating the pointer to the underlying VTK array, and the old pointer is not discarded. The solution to this is to simply reuse the underlying pointer.

---

Requires #2770 

### Bonus changes
- Refactored `set_scalars` use of `configure_scalars_mode` as we were calling that twice when setting scalars for custom opacity.

### CODE REVIEW

File: pyvista/core/datasetattributes.py [hunk 1]
```diff
@@ -750,8 +749,24 @@ def _prepare_array(
...
+            # pyvista needs to use the copy of this object rather than wrapping
+            # the array (which leaves a C++ pointer uncollected.
+            if data.VTKObject is not None:
+                # VTK doesn't support strides, therefore we can't directly
+                # point to the underlying object
+                if data.__array_interface__['strides'] is None:
+                    vtk_arr = copy_vtk_array(data.VTKObject, deep=deep_copy)
+                    if isinstance(name, str):
+                        vtk_arr.SetName(name)
+                        return vtk_arr
```
Comment (adeak) [inline pyvista/core/datasetattributes.py:765]: If we're being careful with references, shouldn't we first check if the name is an str, and only call `copy_vtk_array()` when `isinstance(name, str)`? Or, alternatively, can we `return vtk_arr` when the name is unset?

File: pyvista/core/datasetattributes.py [hunk 2]
```diff
@@ -750,8 +749,24 @@ def _prepare_array(
...
+
+        # convert to numpy type
+        if not isinstance(data, np.ndarray):
+            data = np.asarray(data)
```
Comment (adeak) [inline pyvista/core/datasetattributes.py:769]: I think this could just be
```suggestion
        # convert to numpy type if necessary
        data = np.asarray(data)
```
because if the data is already an ndarray then `np.asarray(data) is data`.

File: tests/plotting/test_collection.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    sphere.point_data['data'] = range(sphere.n_points)
+
+    # this might leave a reference behind if we don't property use the pointer
+    # to the vtk array.
+    sphere.point_data['data'] = sphere.point_data['data']
+
+
+def test_pyvistandarray_strides(sphere):
+    sphere['test_scalars'] = sphere.points[:, 2]
+    assert np.allclose(sphere['test_scalars'], sphere.points[:, 2])
```
Comment (adeak) [inline tests/plotting/test_collection.py:15]: I assume this file is meant as a regression test for now-fixed VTK leaks. Is there any way we can verify that no leaks are happening? Or is this already subject to the usual VTK GC checking? If so, did these tests fail before this fix?

File: tests/plotting/test_collection.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+"""This module contains any tests which cause memory leaks."""
+import numpy as np
+
+
+def test_pyvistandarray_assign(sphere):
+    sphere.point_data['data'] = range(sphere.n_points)
+
+    # this might leave a reference behind if we don't property use the pointer
```
Comment (adeak) [inline tests/plotting/test_collection.py:8]:
```suggestion
    # this might leave a reference behind if we don't properly use the pointer
```

File: tests/plotting/test_plotting.py [hunk 5]
```diff
@@ -1647,7 +1647,8 @@ def test_opacity_by_array_uncertainty(uniform):
 
 
 def test_opacity_by_array_user_transform(uniform):
-    uniform['Spatial Point Data'] /= uniform['Spatial Point Data'].max()
+    scaled = uniform['Spatial Point Data'] / uniform['Spatial Point Data'].max()
+    uniform['Spatial Point Data'] = scaled
```
Comment (adeak) [inline tests/plotting/test_plotting.py:1651]: Why was this change necessary?

File: pyvista/utilities/misc.py [hunk 6]
```diff
@@ -63,4 +63,51 @@ def uses_egl() -> bool:
     return 'EGL' in ren_win_str or 'OSOpenGL' in ren_win_str
 
 
+def copy_vtk_array(array, deep=True):
+    """Create a deep or shallow copy a of vtk array.
+
+    Parametres
```
Comment (adeak) [inline pyvista/utilities/misc.py:69]:
```suggestion
    """Create a deep or shallow copy of a VTK array.

    Parameters
```

File: pyvista/utilities/misc.py [hunk 7]
```diff
@@ -63,4 +63,51 @@ def uses_egl() -> bool:
...
+    >>> arr_copy.GetValue(0)
+    1.0
+
+    """
+    if not isinstance(array, (_vtk.vtkDataArray, _vtk.vtkAbstractArray)):
+        raise TypeError(f"Invalid type {type(array)}.")
+
+    vtk_type = str(type(array)).split('.')[-1][:-2]
+    if vtk_type not in dir(_vtk):  # pragma: no cover
+        raise RuntimeError(f'Type {vtk_type} not supported.')
```
Comment (adeak) [inline pyvista/utilities/misc.py:102]: This, along with the superfuous imports in `_vtk` look a bit sketchy. Any chance we can replace all this (up to and including the `getattr()` call) with
```py
new_array = type(array)()
```
or are there edge cases for which we need the specific list of types to be imported?

File: tests/test_utilities.py [hunk 8]
```diff
@@ -778,3 +778,22 @@ def test_has_duplicates():
...
+def test_copy_vtk_array():
+    with pytest.raises(TypeError, match='Invalid type'):
+        pyvista.utilities.misc.copy_vtk_array([1, 2, 3])
+
+    value = 10
+    arr = vtk.vtkFloatArray()
+    arr.SetNumberOfValues(2)
+    arr.SetValue(0, value)
+    arr_copy = pyvista.utilities.misc.copy_vtk_array(arr, deep=True)
+    arr_copy.GetValue(0)
```
Comment (adeak) [inline tests/test_utilities.py:792]: This call seems to "work" even when no value is set. So perhaps this line should be replaced with
```suggestion
    assert arr_copy.GetNumberOfValues()
```
if the failure mode it expects is "the copied array is blank". Or just removed otherwise.

File: tests/test_utilities.py [hunk 9]
```diff
@@ -778,3 +778,22 @@ def test_has_duplicates():
...
+    assert value == arr_copy.GetValue(0)
+
+    arr_copy_not_deep = pyvista.utilities.misc.copy_vtk_array(arr, deep=False)
+    new_value = 5
+    arr.SetValue(1, new_value)
+    arr_copy.GetValue(1)
```
Comment (adeak) [inline tests/test_utilities.py:798]: Similar to the above, not sure what this line tests. The original (deep) copy doesn't have any value set for index 1, so this result would be undefined I think. What we might want to test is that the old deep copy didn't change, but for that we have to initialise index 1 before the first copy.

File: tests/test_utilities.py [hunk 10]
```diff
@@ -778,3 +778,22 @@ def test_has_duplicates():
...
+
+    value = 10
+    arr = vtk.vtkFloatArray()
+    arr.SetNumberOfValues(2)
+    arr.SetValue(0, value)
+    arr_copy = pyvista.utilities.misc.copy_vtk_array(arr, deep=True)
+    arr_copy.GetValue(0)
+    assert value == arr_copy.GetValue(0)
+
+    arr_copy_not_deep = pyvista.utilities.misc.copy_vtk_array(arr, deep=False)
```
Comment (adeak) [inline tests/test_utilities.py:795]: Nitpick, but the fewer indirections, the better:
```suggestion
    arr_copy_shallow = pyvista.utilities.misc.copy_vtk_array(arr, deep=False)
```

File: tests/test_utilities.py [hunk 11]
```diff
@@ -778,3 +778,22 @@ def test_has_duplicates():
...
+    new_value = 5
+    arr.SetValue(1, new_value)
+    arr_copy.GetValue(1)
+    assert new_value == arr_copy_not_deep.GetValue(1)
```
Comment (adeak) [inline tests/test_utilities.py:799]:
```suggestion
    assert new_value == arr_copy_shallow.GetValue(1)
```

File: pyvista/plotting/mapper.py [hunk 12]
```diff
@@ -239,10 +229,38 @@ def configure_scalars_mode(
...
+
+            n_colors : int
+                Number of colors.
+
+            preference : str
+                Either ``'point'`` or ``'cell'``.
+
+            interpolate_before_map : bool
+                Enabling makes for a smoother scalars display.  Default is
+                ``True``.  When ``False``, OpenGL will interpolate the
```
Comment (adeak) [inline pyvista/plotting/mapper.py:255]: This function doesn't have defaults, so this is a bit confusing. Then again this docstring probably doesn't get rendered anywhere.

File: pyvista/plotting/mapper.py [hunk 13]
```diff
@@ -239,10 +229,38 @@ def configure_scalars_mode(
...
+            Parameters
+            ----------
+            scalars : numpy.ndarray
+                Array of scalars to assign to the mapper.
+
+            mesh : pyvista.Dataset
+                Dataset to assign the scalars to.
+
+            title : str
+                Name to assign the scalars.
```
Comment (adeak) [inline pyvista/plotting/mapper.py:245]: Seems like this does two things: if `title` is already the name of a data array then it will be fetched, otherwise the input `scalars` will be added to the mesh with name `title`. If I understand this correctly, this is a bit confusing and might need a mention here.

File: tests/test_utilities.py (hunk 9 above)
Comment (akaszynski) [reply tests/test_utilities.py:799]: Resolved in 129477c3c4867152eb1dfce0f50730aa61c3bc0c.

File: pyvista/core/datasetattributes.py (hunk 1 above)
Comment (akaszynski) [reply pyvista/core/datasetattributes.py]: Opted to returning vtk_arr always and setting the name in a branch.

Resolved in 129477c3c4867152eb1dfce0f50730aa61c3bc0c.

File: pyvista/plotting/mapper.py (hunk 13 above)
Comment (akaszynski) [reply pyvista/plotting/mapper.py]: Agreed.  Resolved in 129477c3c4867152eb1dfce0f50730aa61c3bc0c.

File: pyvista/plotting/mapper.py (hunk 12 above)
Comment (akaszynski) [reply pyvista/plotting/mapper.py]: Documentation clarified in 129477c3c4867152eb1dfce0f50730aa61c3bc0c.

File: pyvista/utilities/misc.py (hunk 7 above)
Comment (akaszynski) [reply pyvista/utilities/misc.py]: This was done primarily for debugging and should have been left out.

Resolved in 129477c3c4867152eb1dfce0f50730aa61c3bc0c.

File: tests/plotting/test_plotting.py (hunk 5 above)
Comment (akaszynski) [reply tests/plotting/test_plotting.py]: This was the most perplexing part of this PR, and I was going to gloss over it until I figured it out, but you're to good for anything to sneak past you.

Accessing an array directly from a dataset actually returns a `numpy.ndarray` rather than a `pyvista.pyvista_ndarray` due to the way that `pyvista.utilities.helpers.point_array` works. For whatever reason, operating on a `numpy.ndarray` representation of our data leaves an uncollected VTK array.

This has been corrected in ea76919e6b7bacc42b659ae86c6de8bc74f42bbd, but no we have to deal with the problem that `pyvista_ndarray.max` returns a one element array that is treated differently than a size 1 `numpy.ndarray`. This was already discussed in #1953 and commented [here](https://github.com/pyvista/pyvista/pull/1953#issuecomment-1146896567).

This will have to be resolved in a separate PR first.

File: tests/plotting/test_collection.py (hunk 3 above)
Comment (akaszynski) [reply tests/plotting/test_collection.py]: All files in the `tests/plotting` directory are subject to our VTK GC checking. These tests failed before this.

File: tests/plotting/test_plotting.py (hunk 5 above)
Comment (adeak) [reply tests/plotting/test_plotting.py]: Flattery will get you nowhere :laughing: For what it's worth I would've been OK with keeping this as-is and postponing its investigation for later.

So if part of the issue was that
> Accessing an array directly from a dataset actually returns a `numpy.ndarray` rather than a `pyvista.pyvista_ndarray` due to the way that `pyvista.utilities.helpers.point_array` works
do we want to change that? Or is that behaviour necessary/expected/etc.?

File: pyvista/_vtk.py [hunk 14]
```diff
@@ -111,16 +111,34 @@
...
+        vtkTypeUInt8Array,
+        vtkTypeUInt16Array,
         vtkTypeUInt32Array,
+        vtkTypeUInt64Array,
+        vtkUnicodeStringArray,
         vtkUnsignedCharArray,
+        vtkUnsignedIntArray,
+        vtkUnsignedLongArray,
+        vtkUnsignedLongLongArray,
+        vtkUnsignedShortArray,
```
Comment (adeak) [inline pyvista/_vtk.py:141]: If we're no longer relying on checking for these new array subclasses in `_vtk`, should we revert these added imports?

File: pyvista/utilities/helpers.py [hunk 15]
```diff
@@ -215,10 +215,15 @@ def point_array(obj, name):
 
     Returns
     -------
-    numpy.ndarray
+    pyvista.pyvista_ndarray or numpy.ndarray
         Wrapped array.
```
Comment (adeak) [inline pyvista/utilities/helpers.py:219]: Should we explain what determines the return type? Assuming we keep this duality.

Same for field and cell arrays below.

File: tests/test_utilities.py (hunk 9 above)
Comment (adeak) [reply tests/test_utilities.py:799]: I think this line survived the cleanup by accident. There's still a throwaway `GetValue(1)` line, instead I'd expect something like `assert value == arr_copy.GetValue(1)`

Comment (akaszynski) [reply tests/test_utilities.py:799]: Quite right.

```suggestion
    assert value == arr_copy.GetValue(1)
```

File: tests/plotting/test_plotting.py (hunk 5 above)
Comment (akaszynski) [reply tests/plotting/test_plotting.py]: I’ve changed it in #2770. We should always return pyvista_ndarray.

File: pyvista/_vtk.py (hunk 14 above)
Comment (akaszynski) [reply pyvista/_vtk.py]: Done.

File: pyvista/utilities/helpers.py (hunk 15 above)
Comment (akaszynski) [reply pyvista/utilities/helpers.py]: Always returning pyvista_ndarray.

**Comment by akaszynski** [comment]
Ready for review @adeak.

File: pyvista/utilities/helpers.py [hunk 16]
```diff
@@ -215,12 +215,16 @@ def point_array(obj, name):
...
-    return convert_array(vtkarr)
+    if isinstance(obj, pyvista.DataSet):
+        try:
+            return obj.point_data.get_array(name)
+        except KeyError:  # pragma: no cover
+            return None
+    return pyvista.pyvista_ndarray(obj.GetPointData().GetAbstractArray(name))
 
 
 def field_array(obj, name):
```
Comment (adeak) [inline pyvista/utilities/helpers.py:230]: Looking at this duplication in the original code and the this PR, how bad would it be to define an internal helper that does something like
```py
def _assoc_array(obj, name, association='point'):
    VTK_attr = f'Get{association.title()}Data'
    python_attr = f'{association.lower()}_data'
    if isinstance(obj, pyvista.DataSet):
        return getattr(obj, python_attr).get(name)
    return pyvista.pyvista_ndarray(getattr(obj, VTK_attr)().GetAbstractArray(name))
```
(potentially with more temporary variables for readability) and then making `point_array`, `field_array` and `cell_array` a thin wrapper for that?

File: pyvista/utilities/helpers.py [hunk 17]
```diff
@@ -215,12 +215,16 @@ def point_array(obj, name):
...
         Wrapped array.
 
     """
-    vtkarr = obj.GetPointData().GetAbstractArray(name)
-    return convert_array(vtkarr)
+    if isinstance(obj, pyvista.DataSet):
+        try:
+            return obj.point_data.get_array(name)
+        except KeyError:  # pragma: no cover
+            return None
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/utilities/helpers.py:226]: Sadly, we need to still permit the usage of `int` as a key. I'd prefer using only string keys, but as we still need to support getting arrays by index, we're stuck with continuing to have to support this.

File: pyvista/utilities/helpers.py [hunk 18]
```diff
@@ -215,12 +215,16 @@ def point_array(obj, name):
 
     Returns
     -------
-    numpy.ndarray
+    pyvista.pyvista_ndarray
         Wrapped array.
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/utilities/helpers.py]: Fixed in ff94f28b164d4b81bf353c524acc414c0c8a23c9.

File: pyvista/utilities/helpers.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (adeak) [reply pyvista/utilities/helpers.py:226]: In that case do we want to document the int case? Or is it a permanent but undocumented legacy feature?

File: pyvista/core/datasetattributes.py [hunk 19]
```diff
@@ -750,8 +749,23 @@ def _prepare_array(
...
+                # VTK doesn't support strides, therefore we can't directly
+                # point to the underlying object
+                if data.__array_interface__['strides'] is None:
+                    vtk_arr = copy_vtk_array(data.VTKObject, deep=deep_copy)
+                    if isinstance(name, str):
+                        vtk_arr.SetName(name)
+                    return vtk_arr
+
+        # convert to numpy type if necessary
+        data = np.asarray(data)
```
Comment (adeak) [inline pyvista/core/datasetattributes.py:768]: Let's allow subclasses here to prevent copies from `pyvista_ndarray`.
```suggestion
        data = np.asanyarray(data)
```

File: pyvista/core/datasetattributes.py [hunk 20]
```diff
@@ -1070,8 +1084,8 @@ def update(self, array_dict: Union[Dict[str, np.ndarray], 'DataSetAttributes']):
             self[name] = array.copy()
 
     def _raise_index_out_of_bounds(self, index: Any):
-        max_index = self.VTKObject.GetNumberOfArrays()
         if isinstance(index, int):
+            max_index = self.VTKObject.GetNumberOfArrays()
             if index < 0 or index >= self.VTKObject.GetNumberOfArrays():
```
Comment (adeak) [inline pyvista/core/datasetattributes.py:1089]: Originally editing to use `max_index`, but might as well make use of operator chaining:
```suggestion
            if not 0 <= index < max_index:
```

File: pyvista/core/datasetattributes.py [hunk 21]
```diff
@@ -1070,8 +1084,8 @@ def update(self, array_dict: Union[Dict[str, np.ndarray], 'DataSetAttributes']):
...
                 raise KeyError(f'Array index ({index}) out of range [0, {max_index}]')
```
Comment (adeak) [inline pyvista/core/datasetattributes.py:1090]: When phrased like this, we should probably make this inclusive:
```suggestion
                raise KeyError(f'Array index ({index}) out of range [0, {max_index - 1}]')
```

File: pyvista/utilities/helpers.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (adeak) [reply pyvista/utilities/helpers.py]: Ah, already added, thanks.

File: pyvista/utilities/helpers.py (hunk 16 above)
Comment (akaszynski) [reply pyvista/utilities/helpers.py:243]: Agree with [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). Refactored in 1d155ca710ef906ec7bb2554f3a5c620b6a0e05d.

File: pyvista/core/datasetattributes.py [hunk 22]
```diff
@@ -750,8 +749,23 @@ def _prepare_array(
...
+
+        # attempt to reuse the existing pointer to underlying VTK data
+        if isinstance(data, pyvista_ndarray):
+            # pyvista_ndarray already contains the reference to the vtk object
+            # pyvista needs to use the copy of this object rather than wrapping
+            # the array (which leaves a C++ pointer uncollected.
+            if data.VTKObject is not None:
+                # VTK doesn't support strides, therefore we can't directly
+                # point to the underlying object
+                if data.__array_interface__['strides'] is None:
```
Comment (adeak) [inline pyvista/core/datasetattributes.py:761]: Is this the same as a check for being c-contiguous? If so, a dunder-free alternative:
```suggestion
                if data.flags.c_contiguous:
```

**Comment by akaszynski** [comment]
@adeak, we're ready for review (again).

File: pyvista/utilities/helpers.py [hunk 23]
```diff
@@ -202,6 +202,27 @@ def is_pyvista_dataset(obj):
...
+
+    if isinstance(obj, pyvista.DataSet):
+        try:
+            return getattr(obj, python_attr).get_array(name)
+        except KeyError:  # pragma: no cover
+            return None
+    abstract_array = getattr(obj, vtk_attr)().GetAbstractArray(name)
+    if abstract_array is not None:
+        return pyvista.pyvista_ndarray(abstract_array)
+    return None
```
Comment (adeak) [inline pyvista/utilities/helpers.py:223]: Coverage was complaining here, and thanks to it I found a bug here. Also added some tests for coverage. Please check.

**Review by adeak — APPROVED** [review]
Approving the earlier changes, but @akaszynski please check mine before merging (assuming CI ends up green).
